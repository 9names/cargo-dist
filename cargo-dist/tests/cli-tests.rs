use std::{
    env::consts::EXE_SUFFIX,
    process::{Command, Output, Stdio},
};

static BIN: &str = env!("CARGO_BIN_EXE_cargo-dist");
const ENV_RUIN_ME: &str = "RUIN_MY_COMPUTER_WITH_INSTALLERS";

#[allow(dead_code)]
mod gallery;
use axoasset::LocalAsset;
use camino::Utf8PathBuf;
use gallery::*;

fn format_outputs(output: &Output) -> String {
    let stdout = std::str::from_utf8(&output.stdout).unwrap();
    let stderr = std::str::from_utf8(&output.stderr).unwrap();
    format!("stdout:\n{stdout}\nstderr:\n{stderr}")
}

#[test]
fn test_version() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("-V")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    let stdout = String::from_utf8(output.stdout).unwrap();
    let stderr = String::from_utf8(output.stderr).unwrap();

    assert!(output.status.success(), "{}", stderr);
    assert_eq!(stderr, "");

    let (name, ver) = stdout.split_once(' ').unwrap();
    assert_eq!(name, "cargo-dist");
    let mut ver_parts = ver.trim().split('.');
    ver_parts.next().unwrap().parse::<u8>().unwrap();
    ver_parts.next().unwrap().parse::<u8>().unwrap();
    let last = ver_parts.next().unwrap();
    if let Some((last, _prerelease)) = last.split_once('-') {
        last.parse::<u8>().unwrap();
        if let Some(build) = ver_parts.next() {
            build.parse::<u8>().unwrap();
        }
    } else {
        last.parse::<u8>().unwrap();
    }
    assert!(ver_parts.next().is_none());
}

#[test]
fn test_long_help() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("--help")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    snapshot_settings().bind(|| {
        insta::assert_snapshot!("long-help", format_outputs(&output));
    });
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_short_help() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("-h")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    snapshot_settings().bind(|| {
        insta::assert_snapshot!("short-help", format_outputs(&output));
    });
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_manifest() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("manifest")
        .arg("--artifacts=lies")
        .arg("--no-local-paths")
        .arg("--allow-dirty")
        .arg("--output-format=json")
        .arg("--verbose=error")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    // We don't want this to churn every time we do a version bump
    snapshot_settings_with_dist_manifest_filter().bind(|| {
        insta::assert_snapshot!(format_outputs(&output));
    });

    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_lib_manifest() {
    let version = std::env!("CARGO_PKG_VERSION");
    let output = Command::new(BIN)
        .arg("dist")
        .arg("manifest")
        .arg("--artifacts=all")
        .arg("--no-local-paths")
        .arg("--allow-dirty")
        .arg("--output-format=json")
        .arg("--verbose=error")
        .arg("--tag")
        .arg(&format!("cargo-dist-schema-v{}", version))
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    // We don't want this to churn every time we do a version bump
    snapshot_settings_with_dist_manifest_filter().bind(|| {
        insta::assert_snapshot!(format_outputs(&output));
    });

    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_lib_manifest_slash() {
    let version = std::env!("CARGO_PKG_VERSION");
    let output = Command::new(BIN)
        .arg("dist")
        .arg("manifest")
        .arg("--artifacts=all")
        .arg("--no-local-paths")
        .arg("--allow-dirty")
        .arg("--output-format=json")
        .arg("--verbose=error")
        .arg("--tag")
        .arg(&format!("cargo-dist-schema/v{}", version))
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    // We don't want this to churn every time we do a version bump
    snapshot_settings_with_dist_manifest_filter().bind(|| {
        insta::assert_snapshot!(format_outputs(&output));
    });

    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_error_manifest() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("manifest")
        .arg("--artifacts=all")
        .arg("--no-local-paths")
        .arg("--allow-dirty")
        .arg("--output-format=json")
        .arg("--verbose=error")
        .arg("--tag=v0.0.0")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    // We don't want this to churn every time we do a version bump
    snapshot_settings_with_dist_manifest_filter().bind(|| {
        insta::assert_snapshot!(format_outputs(&output));
    });

    assert!(!output.status.success(), "{}", output.status);
}

#[test]
fn test_markdown_help() {
    let output = Command::new(BIN)
        .arg("dist")
        .arg("help-markdown")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    snapshot_settings().bind(|| {
        insta::assert_snapshot!("markdown-help", format_outputs(&output));
    });
    assert!(output.status.success(), "{}", output.status);
}

static RECEIPT_TEMPLATE: &str = r#"{"binaries":["cargo-dist"],"install_prefix":"INSTALL_PREFIX","provider":{"source":"cargo-dist","version":"0.10.0-prerelease.1"},"source":{"app_name":"cargo-dist","name":"cargo-dist","owner":"axodotdev","release_type":"github"},"version":"VERSION"}"#;

fn install_receipt(version: &str, prefix: &Utf8PathBuf) -> String {
    RECEIPT_TEMPLATE
        .replace("INSTALL_PREFIX", &prefix.to_string().replace('\\', "\\\\"))
        .replace("VERSION", version)
}

fn write_receipt(version: &str, prefix: &Utf8PathBuf, config_path: &Utf8PathBuf) {
    let contents = install_receipt(version, prefix);
    let receipt_name = config_path.join("cargo-dist-receipt.json");
    LocalAsset::write_new_all(&contents, receipt_name).unwrap();
}

#[test]
fn test_self_update() {
    // Only do this if RUIN_MY_COMPUTER_WITH_INSTALLERS is set
    if std::env::var(ENV_RUIN_ME)
        .map(|s| !s.is_empty())
        .unwrap_or(false)
    {
        let dist_home = Utf8PathBuf::from_path_buf(
            homedir::get_my_home()
                .unwrap()
                .unwrap()
                .join(".cargo")
                .join("bin"),
        )
        .unwrap();
        let dist_path = &dist_home.join(format!("cargo-dist{}", EXE_SUFFIX));

        #[cfg(target_family = "unix")]
        let config_path = Utf8PathBuf::from_path_buf(
            homedir::get_my_home()
                .unwrap()
                .unwrap()
                .join(".config")
                .join("cargo-dist"),
        )
        .unwrap();
        #[cfg(target_family = "windows")]
        let config_path = Utf8PathBuf::from_path_buf(
            std::env::var("LOCALAPPDATA")
                .map(std::path::PathBuf::from)
                .unwrap()
                .join("cargo-dist"),
        )
        .unwrap();

        // Ensure we delete any previous copy that may exist
        // at this path before we copy in our version.
        if dist_path.exists() {
            std::fs::remove_file(dist_path).unwrap();
        }
        assert!(!dist_path.exists());

        // Install to the home directory
        std::fs::copy(BIN, dist_path).unwrap();

        // Create a fake install receipt
        // We lie about being a very old version so we always
        // consider upgrading to something.
        write_receipt("0.5.0", &dist_home, &config_path);

        let output = Command::new(dist_path)
            .arg("dist")
            .arg("update")
            // init includes interactive components, so we
            // can't safely run it within a noninteractive test
            .arg("--skip-init")
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .output()
            .unwrap();

        let out_str = String::from_utf8_lossy(&output.stdout);
        let err_str = String::from_utf8_lossy(&output.stderr);

        assert!(
            output.status.success(),
            "status code: {}, stdout: {out_str}; stderr: {err_str}",
            output.status
        );
    }
}
