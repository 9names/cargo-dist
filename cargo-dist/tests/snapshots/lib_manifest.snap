---
source: cargo-dist/tests/cli-tests.rs
expression: format_outputs(&output)
---
stdout:
{
  "dist_version": "1.0.0-FAKEVERSION",
  "announcement_tag": "CENSORED",
  "announcement_tag_is_implicit": false,
  "announcement_is_prerelease": "CENSORED"
  "announcement_title": "CENSORED"
  "announcement_changelog": "CENSORED"
  "announcement_github_body": "CENSORED"
  "releases": [
    {
      "app_name": "cargo-dist-schema",
      "app_version": "1.0.0-FAKEVERSION",
      "env": {
        "install_dir_env_var": "CARGO_DIST_SCHEMA_INSTALL_DIR",
        "unmanaged_dir_env_var": "CARGO_DIST_SCHEMA_UNMANAGED_INSTALL",
        "disable_update_env_var": "CARGO_DIST_SCHEMA_DISABLE_UPDATE",
        "no_modify_path_env_var": "CARGO_DIST_SCHEMA_NO_MODIFY_PATH",
        "github_base_url_env_var": "CARGO_DIST_SCHEMA_INSTALLER_GITHUB_BASE_URL",
        "github_enterprise_base_url_env_var": "CARGO_DIST_SCHEMA_INSTALLER_GITHUB_ENTERPRISE_BASE_URL"
      },
      "display_name": "cargo-dist-schema",
      "display": true,
      "hosting": {
        "github": {
          "artifact_base_url": "https://github.com",
          "artifact_download_path": "/axodotdev/cargo-dist/releases/download/cargo-dist-schema-v1.0.0-FAKEVERSION",
          "artifact_download_url": "https://github.com/axodotdev/cargo-dist/releases/download/cargo-dist-schema-v1.0.0-FAKEVERSION",
          "owner": "axodotdev",
          "repo": "cargo-dist"
        },
        "axodotdev": {
          "package": "cargo-dist-schema",
          "public_id": "fake-id-do-not-upload",
          "set_download_url": "https://fake.axo.dev/faker/cargo-dist-schema/fake-id-do-not-upload",
          "upload_url": null,
          "release_url": null,
          "announce_url": null
        }
      }
    }
  ],
  "systems": {
    "plan:all:": {
      "id": "plan:all:",
      "cargo_version_line": "CENSORED"
      "build_environment": "indeterminate"
    }
  },
  "publish_prereleases": false,
  "force_latest": false,
  "ci": {
    "github": {
      "artifacts_matrix": {},
      "pr_run_mode": "plan"
    }
  },
  "linkage": [],
  "upload_files": [],
  "github_attestations": true
}

stderr:
INFO: You've enabled Axo Releases, which is currently in Closed Beta.
If you haven't yet signed up, please join our discord
(https://discord.gg/ECnWuUUXQk) or message hello@axo.dev to get started!
